experimental/yarpl/include/yarpl/flowable/FlowableOperator.h

```diff
@@ -133,6 +133,88 @@ class MapOperator : public FlowableOperator<U, D> {
   F function_;
 };
 
+template <
+    typename U,
+    typename D,
+    typename F,
+    typename = typename std::enable_if<std::is_trivially_assignable<D, U>::value>,
+    typename = typename std::enable_if<std::is_callable<F(D, U), D>::value>::type>
+class ReduceOperator : public FlowableOperator<U, D> {
+public:
+  ReduceOperator(Reference<Flowable<U>> upstream, F&& function)
+      : FlowableOperator<U, D>(std::move(upstream)),
+        function_(std::forward<F>(function)) {}
+
+  void subscribe(Reference<Subscriber<D>> subscriber) override {
+    FlowableOperator<U, D>::upstream_->subscribe(
+        // Note: implicit cast to a reference to a subscriber.
+        Reference<Subscription>(new Subscription(
+            Reference<Flowable<D>>(this), std::move(subscriber))));
+  }
+
+private:
+  class Subscription : public FlowableOperator<U, D>::Subscription {
+  public:
+    Subscription(
+        Reference<Flowable<D>> flowable,
+        Reference<Subscriber<D>> subscriber)
+        : FlowableOperator<U, D>::Subscription(
+        std::move(flowable),
+        std::move(subscriber)),
+          accInitialized_(false) {
+    }
+
+    void request(int64_t /* delta */) override {
+      // Request all of the items
+      callSuperRequest(FlowableOperator<U, D>::NO_FLOW_CONTROL);
+    }
+
+    void onNext(U value) override {
+      auto* flowable = FlowableOperator<U, D>::Subscription::flowable_.get();
+      auto* reduce = static_cast<ReduceOperator*>(flowable);
+      if (accInitialized_) {
+        acc_ = reduce->function_(std::move(acc_), std::move(value));
+      } else {
+        acc_ = std::move(value);
+        accInitialized_ = true;
+      }
+    }
+
+    void onComplete() override {
+      auto subscriber =
+          FlowableOperator<U, D>::Subscription::subscriber_.get();
+      if (accInitialized_) {
+        subscriber->onNext(acc_);
+        callSuperOnComplete();
+      } else {
+        callSuperOnError(std::make_exception_ptr(std::runtime_error("Upstream has no value")));
```

Comment:
I thought again and perhaps we should not throw at all and just call complete without calling onNext.
I will do this update along with implementing Observable Reduce Operator unit tests.